annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java

```diff
@@ -60,25 +69,60 @@ public TypeCheck(ConstraintHelper constraintHelper) {
 	}
 
 	@Override
-	public Set<ConstraintCheckIssue> checkNonAnnotationType(
-			TypeElement element, AnnotationMirror annotation) {
-
+	public Set<ConstraintCheckIssue> checkNonAnnotationType(TypeElement element, AnnotationMirror annotation) {
 		return checkInternal( element, annotation, element.asType(), "NOT_SUPPORTED_TYPE" );
 	}
 
-	private Set<ConstraintCheckIssue> checkInternal(Element element,
-			AnnotationMirror annotation, TypeMirror type, String messageKey) {
-
-		if ( constraintHelper.checkConstraint(
-				annotation.getAnnotationType(), type ) != ConstraintCheckResult.ALLOWED ) {
-
+	private Set<ConstraintCheckIssue> checkInternal(Element element, AnnotationMirror annotation, TypeMirror type, String messageKey) {
+		TypeMirror typeToCheck = usesUnwrapping( annotation, type ) ? getUnwrappedType( type ) : type;
+		if ( !isAnnotationAllowedForType( annotation, typeToCheck ) ) {
 			return CollectionHelper.asSet(
 					ConstraintCheckIssue.error(
 							element, annotation, messageKey,
-							annotation.getAnnotationType().asElement().getSimpleName() ) );
+							annotation.getAnnotationType().asElement().getSimpleName()
+					) );
 		}
 
 		return Collections.emptySet();
 	}
 
+	private boolean isAnnotationAllowedForType(AnnotationMirror annotation, TypeMirror type) {
+		return ConstraintCheckResult.ALLOWED.equals( constraintHelper.checkConstraint( annotation.getAnnotationType(), type ) );
+	}
+
+	private boolean usesUnwrapping(AnnotationMirror annotationMirror, TypeMirror typeMirror) {
+		//need to check if this annotation is not on one of the types that are automatically unwrapped:
+		if ( constraintHelper.isSupportedForUnwrappingByDefault( getQualifiedName( typeMirror ) ) ) {
+			return true;
+		}
+
+		//otherwise look for Unwrapping type in payload:
+		return annotationApiHelper.getAnnotationArrayValue( annotationMirror, "payload" ).stream()
+				.map( AnnotationValue::getValue )
+				.map( type -> (TypeMirror) type )
+				.map( typeUtils::asElement )
+				.map( elem -> ( (TypeElement) elem ).getQualifiedName() )
+				.filter( name -> name.toString().equals( "javax.validation.valueextraction.Unwrapping.Unwrap" ) )
+				.findAny().isPresent();
+	}
+
+	private TypeMirror getUnwrappedType(TypeMirror type) {
+		Optional<TypeMirror> optional = constraintHelper.getUnwrappedToByDefault( getQualifiedName( type ) );
+		if ( optional.isPresent() ) {
+			return optional.get();
+		}
+		else {
+			//TODO: need to find a way to check for unwrapping
+			return type;
```

Comment:
By the way, in the case we don't have the unwrapped type, I think we should throw a warning and avoid validating anything after that. Don't know if it's complicated to do that in the current execution flow but I think it's the best we can do.

Note that even if we allow to declare the types, I think this warning thing would still be useful.